storage/country_tree.hpp

```diff
@@ -6,205 +6,234 @@
+        f(*child);
     }
-    return nullptr;
-  }
 
-  bool HasParent() const { return m_parent != nullptr; }
-
-  CountryTree<T> const & Parent() const
-  {
-    CHECK(HasParent(), ());
-    return *m_parent;
-  }
+    template <class TFunctor>
+    void ForEachChild(TFunctor && f) const
+    {
+      for (auto const & child : m_children)
+        f(*child);
+    }
 
-  CountryTree<T> const & Child(size_t index) const
-  {
-    ASSERT_LESS(index, m_children.size(), ());
-    return *m_children[index];
-  }
+    /// \brief Calls functor f for all nodes (add descendant) in the tree.
+    template <class TFunctor>
+    void ForEachDescendant(TFunctor && f)
+    {
+      for (auto & child : m_children)
+      {
+        f(*child);
+        child->ForEachDescendant(f);
+      }
+    }
 
-  size_t ChildrenCount() const
-  {
-    return m_children.size();
-  }
+    template <class TFunctor>
+    void ForEachDescendant(TFunctor && f) const
+    {
+      for (auto const & child: m_children)
+      {
+        f(*child);
+        child->ForEachDescendant(f);
+      }
+    }
 
-  /// \brief Calls functor f for each first generation descendant of the node.
-  template <class TFunctor>
-  void ForEachChild(TFunctor && f)
-  {
-    for (auto & child : m_children)
-      f(*child);
-  }
+    template <class TFunctor>
+    void ForEachInSubtree(TFunctor && f)
+    {
+      f(*this);
+      for (auto & child: m_children)
+        child->ForEachInSubtree(f);
+    }
 
-  template <class TFunctor>
-  void ForEachChild(TFunctor && f) const
-  {
-    for (auto const & child : m_children)
-      f(*child);
-  }
+    template <class TFunctor>
+    void ForEachInSubtree(TFunctor && f) const
+    {
+      f(*this);
+      for (auto const & child: m_children)
+        child->ForEachInSubtree(f);
+    }
 
-  /// \brief Calls functor f for all nodes (add descendant) in the tree.
-  template <class TFunctor>
-  void ForEachDescendant(TFunctor && f)
-  {
-    for (auto & child : m_children)
+    template <class TFunctor>
+    void ForEachAncestorExceptForTheRoot(TFunctor && f)
     {
-      f(*child);
-      child->ForEachDescendant(f);
+      if (m_parent == nullptr || m_parent->m_parent == nullptr)
+        return;
+      f(*m_parent);
+      m_parent->ForEachAncestorExceptForTheRoot(f);
     }
-  }
 
-  template <class TFunctor>
-  void ForEachDescendant(TFunctor && f) const
-  {
-    for (auto const & child: m_children)
+    template <class TFunctor>
+    void ForEachAncestorExceptForTheRoot(TFunctor && f) const
     {
-      f(*child);
-      child->ForEachDescendant(f);
+      if (m_parent == nullptr || m_parent->m_parent == nullptr)
+        return;
+      f(*m_parent);
+      m_parent->ForEachAncestorExceptForTheRoot(f);
     }
-  }
+  };
+
+private:
+  using TCountryTreeHashTable = unordered_multimap<I, Node *, NodeKeyHasher<I>>;
+public:
+
+  CountryTree(T const & value = T(), Node * parent = nullptr)
+    : m_countryTree(value, parent) {}
```

Comment:
Is it ok that the tree can not be empty? The usual approach is to use unique_ptr<Node> m_countryTree instead of Node m_countryTree.